From 9ac598a12130b09a16f413c0451036bfd7343091 Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:20:53 -0400
Subject: [alsa] hda: Avoid codec D3 for keeping mute LED up on Lenovo Yxx0

Message-id: <140972823573-136-patchset-jkysela@redhat.com>
Patchwork-id: 90836
O-Subject: [RHEL 7.1 PATCH 135/442] ALSA: hda - Avoid codec D3 for keeping mute LED up on Lenovo Yxx0
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

The GPIO line used for the mute LED control on Lenovo Yxx0 laptops is
cleared unexpectedly when the codec goes to D3, typically by
power-saving.  For avoiding it, add a power filter in the fixup.

Bugzilla: https://bugzilla.kernel.org/show_bug.cgi?id=16373
Signed-off-by: Takashi Iwai <tiwai@suse.de>

(picked from kernel: bf68665d7a5647591258c120382fd64465db7d63)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/sound/pci/hda/patch_realtek.c b/sound/pci/hda/patch_realtek.c
index 28cdbb8..ee1ba22 100644
--- a/sound/pci/hda/patch_realtek.c
+++ b/sound/pci/hda/patch_realtek.c
@@ -4940,6 +4940,17 @@ static void alc662_led_gpio1_mute_hook(void *private_data, int enabled)
         spec->gpio_led);
 }
 
+/* avoid D3 for keeping GPIO up */
+static unsigned int gpio_led_power_filter(struct hda_codec *codec,
+       hda_nid_t nid,
+       unsigned int power_state)
+{
+ struct alc_spec *spec = codec->spec;
+ if (nid == codec->afg && power_state == AC_PWRST_D3 && spec->gpio_led)
+  return AC_PWRST_D0;
+ return power_state;
+}
+
 static void alc662_fixup_led_gpio1(struct hda_codec *codec,
        const struct hda_fixup *fix, int action)
 {
@@ -4954,6 +4965,7 @@ static void alc662_fixup_led_gpio1(struct hda_codec *codec,
   spec->gen.vmaster_mute.hook = alc662_led_gpio1_mute_hook;
   spec->gpio_led = 0;
   snd_hda_add_verbs(codec, gpio_init);
+  codec->power_filter = gpio_led_power_filter;
  }
 }
 
-- 
1.7.1